### PR TITLE
Feature/fm 338 da spreadsheet deliverable matrix volume

### PR DESCRIPTION
**Pull request by tariqhamid**

### CODE REVIEW

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb [hunk 1]
```diff
@@ -2,11 +2,20 @@
...
-  def initialize(building_ids_with_service_codes)
+  def initialize(building_ids_with_service_codes, buildings = nil, uvals = nil)
     @building_ids_with_service_codes = building_ids_with_service_codes
     building_ids = building_ids_with_service_codes.map { |h| h[:building_id] }
-    @buildings = FacilitiesManagement::Buildings.where(id: building_ids)
-    buildings_with_service_codes
+
+    @uvals = uvals
+
+    @buildings = buildings || FacilitiesManagement::Buildings.where(id: building_ids)
```
Comment (kennyevil) [inline app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: As discussed, we don't need to pass the building objects in here. We pass in the building ids and use those to call up the buildings from the database.

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb [hunk 2]
```diff
@@ -189,4 +215,46 @@ def add_service_matrix(sheet)
...
+      # add_service_matrix(sheet)
+      # style_service_matrix_sheet(sheet, standard_column_style)
+
+      codes = @building_ids_with_service_codes.collect do |building|
+        building[:service_codes]
+      end
+
+      codes.flatten!.uniq!
+
+      services = FacilitiesManagement::StaticData.work_packages.map { |w| [w['code'], w] }.to_h
```
Comment (kennyevil) [inline app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: ```set_services``` already created an object called ```@services``` and ordered them already. This is duplicated effort.

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb [hunk 3]
```diff
@@ -189,4 +215,46 @@ def add_service_matrix(sheet)
...
+          suv = uvs.find { |u| s == u[:service_code] }
+
+          title_text = suv[:title_text]
+          spreadsheet_label = suv[:spreadsheet_label]
+
+          new_row << suv[:uom_value] if suv
+          new_row << nil unless suv
+        end
+
+        new_row = ([s, services[s]['description']] << title_text << spreadsheet_label << new_row).flatten
```
Comment (kennyevil) [inline app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: Set ```new_row``` above line 241 with the service code, service name, title_text and spreadsheet_label. Then you can shovel in the ```uom_value``` for the buildings that need to provide them.

Comment (tariqhamid) [reply app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: good spot

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb (hunk 2 above)
Comment (tariqhamid) [reply app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: the 'services' defined above are a had keyed on service

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb (hunk 3 above)
Comment (tariqhamid) [reply app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: There's a subtlety in the way title_text are set spreadsheet_label and used.

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb (hunk 1 above)
Comment (tariqhamid) [reply app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: This has been 'fixed'.

We should be using more IoC.

File: app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb (hunk 2 above)
Comment (tariqhamid) [reply app/services/facilities_management/deliverable_matrix_spreadsheet_creator.rb]: fixed

**Pull request merged**